messaging/src/main/java/org/springframework/security/messaging/context/SecurityContextPropagationChannelInterceptor.java

```diff
@@ -0,0 +1,164 @@
+ * in the {@link #beforeHandle(Message, MessageChannel, MessageHandler)} and
+ * {@link #postReceive(Message, MessageChannel)} both of which typically happen on a
+ * different thread.
+ * <p>
+ * Note: cannot be used in combination with a {@link SecurityContextChannelInterceptor} on
+ * the same channel since both these interceptors modify a security context on a handling
+ * and receiving operations.
+ *
+ * @author Artem Bilan
+ * @since 6.2
+ * @see SecurityContextChannelInterceptor
+ */
+public final class SecurityContextPropagationChannelInterceptor implements ExecutorChannelInterceptor {
+
+	private static final ThreadLocal<Stack<SecurityContext>> originalContext = new ThreadLocal<>();
+
+	private SecurityContextHolderStrategy securityContextHolderStrategy = SecurityContextHolder
+			.getContextHolderStrategy();
+
+	private SecurityContext empty = this.securityContextHolderStrategy.createEmptyContext();
+
+	private final String authenticationHeaderName;
+
+	/**
+	 * Create a new instance using the header of the name
+	 * {@link SimpMessageHeaderAccessor#USER_HEADER}.
+	 */
+	public SecurityContextPropagationChannelInterceptor() {
+		this(SimpMessageHeaderAccessor.USER_HEADER);
+	}
+
+	/**
+	 * Create a new instance that uses the specified header to populate the
+	 * {@link Authentication}.
+	 * @param authenticationHeaderName the header name to populate the
+	 * {@link Authentication}. Cannot be null.
+	 */
+	public SecurityContextPropagationChannelInterceptor(String authenticationHeaderName) {
+		Assert.notNull(authenticationHeaderName, "authenticationHeaderName cannot be null");
+		this.authenticationHeaderName = authenticationHeaderName;
+	}
+
+	public void setSecurityContextHolderStrategy(SecurityContextHolderStrategy strategy) {
+		this.securityContextHolderStrategy = strategy;
+		this.empty = this.securityContextHolderStrategy.createEmptyContext();
+	}
+
+	@Override
+	public Message<?> preSend(Message<?> message, MessageChannel channel) {
+		Authentication authentication = this.securityContextHolderStrategy.getContext().getAuthentication();
+		if (authentication != null) {
+			return MessageBuilder.fromMessage(message).setHeader(this.authenticationHeaderName, authentication).build();
+		}
+		return message;
+	}
+
+	@Override
+	public Message<?> beforeHandle(Message<?> message, MessageChannel channel, MessageHandler handler) {
+		return postReceive(message, channel);
+	}
+
+	@Override
+	public Message<?> postReceive(Message<?> message, MessageChannel channel) {
+		setup(message);
+		return message;
+	}
+
+	@Override
+	public void afterMessageHandled(Message<?> message, MessageChannel channel, MessageHandler handler, Exception ex) {
+		cleanup(message);
+	}
+
+	private void setup(Message<?> message) {
+		Authentication authentication = message.getHeaders().get(this.authenticationHeaderName, Authentication.class);
+		if (authentication != null) {
```

Comment:
Instead of conditionally adding to the stack, can you use the null object pattern that `SecurityContextChannelInterceptor` does?

Reply:
In addition to removing the if statement, it also aligns well with the Spring Security Filter Chain which adds an anonymous authentication as well.

Reply:
This interceptor is not about authentication, but rather accepting existing auth from message header or ignoring otherwise.
If you would like to have an anonymous auth, then use a `SecurityContextChannelInterceptor` in the flow after this propagation has done its "void" logic.
Plus it doesn't look very optimal to set a context where it was not requested to be set.

Reply:
My main concern is conditionally adding and removing from a stack. It's an opportunity for memory leaks. 

Consider, for example, if the authentication changes between `postReceive` and `afterMessageHandled`. Then, the stack will never get cleaned up.

I don't think that comparing authentication instances is a reliable way to ensure the stack values get popped correctly.

The check performed at cleanup is trying to figure out if something was placed on the stack so that it can be removed. Instead, having a placeholder value would make the logic simpler and less likely to experience a memory leak.

Would adding an anonymous value here (I see this as meaning "no authentication was propagated") prevent you from being able to use the interceptor or make it difficult?

Reply:
The `afterMessageHandled()` is not called after `postReceive()`. If that is the case, then another instance of the interceptor has to be used downstream. And that's exactly why I have introduced that `public cleanup(Message)`. There is no connection between received message from the `PollableChannel` and downstream handler. But the point is that auth has to be populated to the context to make that handler downstream called in a secure way.

Reply:
OK. I'll look today into your anonymous suggestion.
No, it won't make the interceptor not working since the point is to make this as transparent for Messaging as possible and fully rely on existing features in Spring Security. Therefore if anonymous is OK to reject the call for `@PreAuthorize`, I'm fine to go ahead. Only my concern that it is sub-optimal to always populate into the stack whatever was not requested.

Reply:
Another note: with an anonymous support in this interceptor it is going to be very similar to what we have so far in the `SecurityContextChannelInterceptor`. And therefore we slowly coming back to whatever I suggested to do in my original commit for this PR.

Reply:
I see your point here about them being similar in many ways.

I think they are fundamentally different since `preSend` wants to do the opposite thing in each. The nice thing is we can always combine them in an abstract class or a proper subclass if that's what it comes down to. I don't mind people declaring that their intent is to authenticate vs propagate.